invokeai/frontend/web/src/features/nodes/util/graphBuilders/addHrfToGraph.ts

```diff
@@ -0,0 +1,210 @@
+  METADATA_ACCUMULATOR,
+  LATENTS_TO_IMAGE_HRF,
+  DENOISE_LATENTS_HRF,
+  RESCALE_LATENTS,
+  NOISE_HRF,
+  VAE_LOADER,
+} from './constants';
+
+// Copy certain connections from previous DENOISE_LATENTS to new DENOISE_LATENTS_HRF.
+function copyConnectionsToDenoiseLatentsHrf(graph: NonNullableGraph): void {
+  const destinationFields = [
+    'vae',
+    'control',
+    'ip_adapter',
+    'metadata',
+    'unet',
+    'positive_conditioning',
+    'negative_conditioning',
+  ];
+  const newEdges: Edge[] = [];
+
+  // Loop through the existing edges connected to DENOISE_LATENTS
+  graph.edges.forEach((edge: Edge) => {
+    if (
+      edge.destination.node_id === DENOISE_LATENTS &&
+      destinationFields.includes(edge.destination.field)
+    ) {
+      // Add a similar connection to DENOISE_LATENTS_HRF
+      newEdges.push({
+        source: {
+          node_id: edge.source.node_id,
+          field: edge.source.field,
+        },
+        destination: {
+          node_id: DENOISE_LATENTS_HRF,
+          field: edge.destination.field,
+        },
+      });
+    }
+  });
+  graph.edges = graph.edges.concat(newEdges);
+}
+
+// Adds the high-res fix feature to the given graph.
+export const addHrfToGraph = (
+  state: RootState,
+  graph: NonNullableGraph
+): void => {
+  const { vae } = state.generation;
+  const isAutoVae = !vae;
+
+  // Pre-existing (original) graph nodes.
+  const originalDenoiseLatentsNode = graph.nodes[
+    DENOISE_LATENTS
+  ] as DenoiseLatentsInvocation;
+  const originalNoiseNode = graph.nodes[NOISE] as NoiseInvocation;
+  const originalLatentsToImageNode = graph.nodes[
+    LATENTS_TO_IMAGE
+  ] as LatentsToImageInvocation;
+
+  // Scale height and width by hrfScale.
+  const hrfScale = state.generation.hrfScale;
+  const scaledHeight = originalNoiseNode?.height
+    ? originalNoiseNode.height * hrfScale
+    : undefined;
+  const scaledWidth = originalNoiseNode?.width
+    ? originalNoiseNode.width * hrfScale
+    : undefined;
+
+  // Add hrf information to the metadata accumulator.
+  const metadataAccumulator = graph.nodes[METADATA_ACCUMULATOR] as
+    | MetadataAccumulatorInvocation
+    | undefined;
+  if (metadataAccumulator) {
+    metadataAccumulator.height = scaledHeight;
+    metadataAccumulator.width = scaledWidth;
+  }
+
+  // Define new nodes.
+  // Denoise latents node to be run on upscaled latents.
+  const denoiseLatentsHrfNode: DenoiseLatentsInvocation = {
+    type: 'denoise_latents',
+    id: DENOISE_LATENTS_HRF,
+    is_intermediate: originalDenoiseLatentsNode?.is_intermediate,
+    cfg_scale: originalDenoiseLatentsNode?.cfg_scale,
+    scheduler: originalDenoiseLatentsNode?.scheduler,
+    steps: originalDenoiseLatentsNode?.steps,
+    denoising_start: 1 - state.generation.hrfStrength,
+    denoising_end: 1,
+  };
+
+  const rescaleLatentsNode: RescaleLatentsInvocation = {
+    id: RESCALE_LATENTS,
+    type: 'lresize',
+    width: scaledWidth,
+    height: scaledHeight,
+  };
```

Comment:
As mentioned on discord, this could be an esrgan x4 node and then a resize image node after, but I agree that would be good as followup.